services/venice-controller/src/main/java/com/linkedin/venice/controller/init/ControllerClientBackedSystemSchemaInitializer.java

```diff
@@ -47,23 +58,52 @@ public ControllerClientBackedSystemSchemaInitializer(
       Schema keySchema,
       UpdateStoreQueryParams storeMetadataUpdate,
       boolean autoRegisterPartialUpdateSchema,
-      boolean enforceSSLOnly) {
+      VeniceControllerConfig controllerConfig) {
     this.protocolDefinition = protocolDefinition;
     this.clusterName = systemStoreCluster;
     this.admin = admin;
     this.keySchema = keySchema;
     this.storeMetadataUpdate = storeMetadataUpdate;
     this.autoRegisterPartialUpdateSchema = autoRegisterPartialUpdateSchema;
-    this.enforceSSLOnly = enforceSSLOnly;
+    this.controllerConfig = controllerConfig;
   }
 
   public void execute() {
     String storeName = protocolDefinition.getSystemStoreName();
     Map<Integer, Schema> schemasInLocalResources = Utils.getAllSchemasFromResources(protocolDefinition);
 
-    String leaderControllerUrl = admin.getLeaderController(clusterName).getUrl(enforceSSLOnly);
-    controllerClient =
-        ControllerClient.constructClusterControllerClient(clusterName, leaderControllerUrl, admin.getSslFactory());
+    String regionName = controllerConfig.getRegionName();
+    if (!controllerConfig.getChildControllerUrl(regionName).isEmpty()) {
+      controllerClient = ControllerClient.constructClusterControllerClient(
+          clusterName,
+          controllerConfig.getChildControllerUrl(regionName),
+          admin.getSslFactory());
+    } else if (!controllerConfig.getChildControllerD2ServiceName().isEmpty()
+        && !controllerConfig.getChildControllerD2ZkHost(regionName).isEmpty()) {
+      controllerClient = new D2ControllerClient(
+          controllerConfig.getChildControllerD2ServiceName(),
+          clusterName,
+          controllerConfig.getChildControllerD2ZkHost(regionName),
+          controllerConfig.isControllerEnforceSSLOnly() ? admin.getSslFactory() : Optional.empty());
+    } else {
+      LOGGER.info("No child controller url or d2 configs. Skip system schema registration via controller client.");
+      return;
+    }
+
+    try {
+      RetryUtils.executeWithMaxAttempt(
+          () -> controllerClient.getLeaderControllerUrl(),
+          3,
+          Duration.ofSeconds(2),
+          Collections.singletonList(Exception.class));
+    } catch (Exception e) {
+      LOGGER.info(
+          "Could not find leader controller after retries. This controller is very likely the first one in "
```

Comment:
Yes, this means the region does not have any live controller yet (current controller is spinning up but is not live either). Otherwise, helix will assign leadership and there will be a leader. Moved the logic into a function.